### PR TITLE
Use batched operations for PowerSGD

### PR DESCRIPTION
**Pull request by MagiaSN**
This implements method proposed in #74907

### CODE REVIEW

**Comment by wayi1** [comment]
> cc @wayi1 

Thanks for cc'ing me. On travel this week. If it's not urgent, I can take a look next week.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 1]
```diff
@@ -15,9 +16,10 @@ def _orthogonalize(matrix, epsilon=0):
     Decide between Gram-Schmidt or QR factorization to orthogonalize the matrix.
```
Comment (wayi1) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:16]: s/matrix/a batch of matrices.

Comment (wayi1) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:16]: Also better to rename "matrix" as "matrices" in L13.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 2]
```diff
@@ -431,26 +433,40 @@ def powerSGD_hook(
             total_Qs_size, device=device, dtype=dtype
         )
 
+    # Batch tensors to compress by shape.
+    shape_to_tensors: Dict[torch.Size, List[torch.Tensor]] = {}
```
Comment (wayi1) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:437]: Optional: you can just use `shape_to_tensors = defaultdict(list)` to save L439-L440.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 3]
```diff
@@ -15,9 +16,10 @@ def _orthogonalize(matrix, epsilon=0):
     Decide between Gram-Schmidt or QR factorization to orthogonalize the matrix.
     QR factorization doesn't work with half-precision, but it is usually faster with a rank > 2.
     """
-    assert len(matrix.shape) == 2 and matrix.shape[1] <= matrix.shape[0]
+    assert len(matrix.shape) == 3 and matrix.shape[2] <= matrix.shape[1]
 
-    rank = matrix.shape[1]
+    batch_size = matrix.shape[0]
```
Comment (wayi1) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:21]: Optional: what about replacing `batch_size` with `num_matrices`?

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 4]
```diff
@@ -431,26 +433,40 @@ def powerSGD_hook(
...
-        n, m = tensor.shape
+    for shape, tensors in shape_to_tensors.items():
+        batch_size = len(tensors)
+        n, m = shape
         matrix_approximation_rank = min(n, m, state.matrix_approximation_rank)
+        if batch_size == 1:
+            # Use the original tensor to avoid copy.
+            batched_tensors_to_compress.append(tensors[0].unsqueeze(0))
+        else:
+            batched_tensors_to_compress.append(torch.stack(tensors))
```
Comment (wayi1) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:457]: I guess will incur extra memory footprint and increase the peak memory usage. Shall we make such batching optimization configurable?

My concern is that for some memory-intensive use cases, enabling such optimization may cause OOM. Wdyt?

**Review by wayi1 — CHANGES_REQUESTED** [review]
Thanks for this great optimization! Most are minor nitpicks. The only concern is the extra memory usage due to stacking tensors of the same shape.

**Comment by MagiaSN** [comment]
> Thanks for this great optimization! Most are minor nitpicks. The only concern is the extra memory usage due to stacking tensors of the same shape.

Thanks for your advice, The memory issue is a concern indeed, I think the best solution is putting tensors of the same shape together in the bucketing stage, so we can replace `stack` with `view`, which introduces no extra memory usage, but this would introduce far more modification than expected, so I add a switch to this optimization and it's turned off by default.

BTW, I see CI falling but didn't figure out how it is related to my update. Anyone have a look?

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 5]
```diff
@@ -327,6 +330,10 @@ def powerSGD_hook(
         bucket (dist.GradBucket): Bucket that stores a 1D flattened gradient tensor that batches multiple per-variable tensors.
             Note that since DDP comm hook only supports single process single device mode,
             only exactly one tensor is stored in this bucket.
+        batch_tensors_with_same_shape (bool): A boolean value that when set to ``True``, compress and decompress tensors with
```
Comment (wayi1) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:333]: This needs to be placed inside `PowerSGDState`. A valid hook cannot accept a third arg.

Comment (wayi1) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:333]: Nit: Replace "A boolean value that when set to ``True``," by "Whether to".

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 6]
```diff
@@ -327,6 +330,10 @@ def powerSGD_hook(
...
+            same shape in a batched operation to achieve higher parallelism. Note that you should also increase the bucket size
+            to make more same-shaped tensors appear in the same bucket, however this may reduce the overlap between computation
+            and communication. Set to ``True`` if the compression / decompression computation is a bottleneck.
```
Comment (wayi1) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:336]: Please add "and increase the memory footprint due to stacking the tensors of the same shape" as another potential downside.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 7]
```diff
@@ -327,6 +330,10 @@ def powerSGD_hook(
         bucket (dist.GradBucket): Bucket that stores a 1D flattened gradient tensor that batches multiple per-variable tensors.
             Note that since DDP comm hook only supports single process single device mode,
             only exactly one tensor is stored in this bucket.
+        batch_tensors_with_same_shape (bool): A boolean value that when set to ``True``, compress and decompress tensors with
+            same shape in a batched operation to achieve higher parallelism. Note that you should also increase the bucket size
```
Comment (wayi1) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:334]: Nit: Use "bucket size (i.e., ``bucket_cap_mb`` arg in DDP ctor)" to be more clear.

**Review by wayi1 — APPROVED** [review]
Thanks for the optimization!

cc: @rohan-varma, @pritamdamania  Do you want to enable `batch_tensors_with_same_shape` arg for some internal use cases to see if there can be any speedup? Some NLP use cases may benefit from this optimization.

**Review by rohan-varma — APPROVED** [review]
Thanks for this change! And thanks @wayi1 for reviewing the changes. 

We will give this a try internally soon, and report back with our findings. It may take some effort to identify a workload with neighbouring tensors of the same shape, but if we find something like that hopefully the perf improvement can be sizeable.

**Review by rohan-varma — CHANGES_REQUESTED** [review]
Looks like lint is failing, can we fix that: https://github.com/pytorch/pytorch/runs/5990406716?check_suite_focus=true? 

Also, would be great to add unittests - we can probably reuse the existing powerSGD tests in https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_nccl.py, and then run all of them with this new option to ensure correctness and parity (correctness wise) with the case when this option is not enabled.

**Comment by MagiaSN** [comment]
The lint error seems to be introduced by rebase, it is not included in the change, but I could fix it together. Also added unittest :)

For the performance test, I suggest using a deep model (with more than 24 layers) and increase `bucket_cap_mb` to the model size, the gradients of each layer are same-shaped, so they can naturally form a batch.

**Review by rohan-varma — APPROVED** [review]
lgtm

**Comment by albanD** [comment]
@pytorchbot revert this please 

See https://github.com/pytorch/pytorch/runs/6058704725?check_suite_focus=true for example failure.

**Comment by MagiaSN** [comment]
Sorry, it seems this branch is behind `master` and some tests are missing. Could someone give suggestions on how to fix this?

```
======================================================================
FAIL [1.175s]: test_correct_module_names (__main__.TestPublicBindings)
An API is considered public, if  its  `__module__` starts with `torch.`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_public_bindings.py", line 350, in test_correct_module_names
    self.assertTrue(not failure_list, msg)
AssertionError: False is not true : Following new APIs ( displayed in the form (module, element, element module) ) were added that do not meet our guidelines for public API Please review  https://docs.google.com/document/d/10yx2-4gs0gTMOimVS403MnoAWkqitS8TUHX73PN8EjE/edit?pli=1# for more information:
('torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook', 'Dict', 'typing')
('torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook', 'defaultdict', 'collections')
```

**Comment by albanD** [comment]
You can rebase on top of master to make sure the test is there.
For that particular error, you can see the doc linked or the wiki: https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation

**Comment by MagiaSN** [comment]
Thanks for the guidance, but this pull request is already closed and I can't add more commits. Tell me if you need a fresh new pull request, or there is something else I could help :)

**Comment by albanD** [comment]
Opening a fresh new PR is the simplest way to go yes.

**Pull request closed without merging**